Proton development

Fenominall opened the pull request:
Developing a test todo app for scrinning

Fenominall committed:
Define 'TodoFeedLoader' protocol and 'TodoItem' model

'TodoFeedLoader' will be loading TodoItems from the remote and local api

Fenominall committed:
Define protocols for caching and image loading

- Added 'TodoFeedCache' protocol for saving Todo items feed.
- Added 'TodoImageCache' protocol for saving image data associated with Todo items.
- Introduced 'TodoItemSaveable' protocol for saving individual Todo items.
- Created 'TodoImageLoader' protocol for loading image data from URLs.

Fenominall committed:
Define 'HTTPClient' protocol for API client boundary

- Created 'HTTPClient' protocol to handle "GET" requests.
- This protocol serves as a boundary for the API client implementation.

Fenominall committed:
Changed protocols and model privacy level to be able to work with them from another module

Fenominall committed:
Created 'URLSessionHTTPClient' class that conforms to the 'HTTPClient' and will be doing "GET" requests with the URLSession API.
Handling Errors and for the correct response type as HTTPURLResponse

Fenominall committed:
Add 'TodoFeedItemsMapper' class for mapping data and response into an array of [TodoItem] and HTTPURLResponse extension for status code validation from (200..<300)

Fenominall committed:
Added 'TodoImageDataMapper' for response and data validation

Fenominall committed:
Define 'LocalTodoItem' struct and 'TodoFeedStore' protocol for local feed storage
'TodoFeedStore' will be acting as a boundary to communicate with the 'LocalFeedLoader' and get the implementation of the 'TodoFeedStore' with the CoreData api for cahcing

Fenominall committed:
Defined 'LocalTodoFeedManager' service for the chacning operations

Make 'LocalTodoFeedManager' conform to the 'TodoFeedLoader' to be able to retrieve the cached feed from the store
Defined convertToModels() in the Array extension to be able to transform '[LocalTodoitem]' into '[TodoItem]'

Fenominall committed:
Fixed typo in the 'TodoFeedStore' for update metod to update only single item

Fenominall committed:
Add saving functionality for 'LocalTodoFeedManager'

- Implemented TodoFeedCache conformance to save a feed of TodoItems to cache.
- Added TodoItemSaveable conformance to save individual TodoItems to cache.
- Introduced helper methods to convert TodoItems to LocalTodoItems for storage.
- Ensured data is correctly saved by converting the models between their local and domain representations.

Renamed the converting methods to clarify the intent adding ed

Fenominall committed:
Implement 'CoreDataFeedStore' with NSPersistentContainer setup, context management, and async support for Core Data operations

Fenominall committed:
Make 'CoreDataFeedStore' conform to the 'TodoFeedStore'
The implementation will be provided later

Fenominall committed:
Added new iOS target to the project

Defined 'AppImageConstants' enum that will hold the name for the system images

Fenominall committed:
Added Traged for the UILayer
Introduced TodoItemViewModel, FeedViewModel that will hold an array of TodoItemViewModel

Did Create necessary views for the TaskListView

Fenominall committed:
Added TaskDetailView and all its components with TaskDetailImageView and DownloadImageButton

Introduced 'TodoItemDetailViewModel'
Added a TodoFeedViewModel DTO for the 'TodoItem'

Refactored TaskListView
Changed the privacy level from properties and object

Fenominall committed:
Created 'TaskRowImageView' that will represent the image or gray rectangle if the image was not provided

Fenominall committed:
Did created App ProtonTodoDevTestApp for the composition

Created genric LoadResourcePresentationAdapter that will trigger the API calls and returning the result
Defined ListViewComposer for composing TaskListView

Add getPublisher method to HTTPClient for async URL fetching and caching extension for Publisher

Fenominall committed:
Did make 'CoreDataFeedStore' public to be able to Instantiate it and its necessary methods from another target
Did create 'NullStore' object taht conform to the 'TodoFeedStore' and will be returned when we cannot instantiate CoreDataFeedStore to avoid crash

Fenominall committed:
Make 'LocalTofoFeedManager' init public
Make 'TodoFeedItemsMapper' static func map as public
Added caching extension and loadPublisher extension for the LocalTodoFeedManager in CombineHelpers
Introduced

renamed ListViewComposer to TaskListViewComposer

Add fallback mechanism for Publishers

- Introduced an extension on Publisher to provide a fallback mechanism
- If the primary publisher fails, it switches to a fallback publisher
- Ensures resilience in data fetching and improves error handling

Instantiated LocalTodoFeedManager in the entry point of the app

Fenominall committed:
Added CoreDataModel with .xcdatamodeled as a TodoFeedStore Provided the model model name and path in the CoreDataFeedStore

Fenominall committed:
Removed unused code

Fenominall committed:
-Provided implemntation for the 'LocalTodoImageCachingManager' with the' TodoImageLoader' and 'TodoImageCache' protocol
-Introduced 'TodoFeedImageStore'that was implented by the 'CoreDataFeedStore'
-Changed the previacy level of the map func to public in 'TodoFeedImageManager'
- Handeling more error edge cases in the 'URLSessionHTTPClient'
- Removed unused method from the Combine helpers saveIgnoringResult
- Introduced new Combine helper Publishers -  loadImageDataPublisher and caching for the 'TodoImageLoader' and 'TodoImageCache'
- Make Null Store conform to the 'TodoFeedImageStore' and provde no implementation
- Added helper metods in the enry point of the app for api request requests to the local database with the fallback mechanism.
if API fails we try to retrieve from the cache and vise versa, also with the mapping logic

Fenominall committed:
Added ThreadSafeDictionary for safe muthreading operations with the dictionary

Fenominall committed:
Feature changes:

'TaskListview' changes:
    - Wrapped TaskRowView in the button to be able provide a necessary action, potenially I will be passing task.id to find the TodoItem
    - added .refreshable and onAppear with the viewModel.load() to the view for feed loading

'TaskListViewComposer' changes:
    - Defined 'LoadResourcePresentationAdapter' implementation where I trigger load from the viewModels closure
    - Changed the selection parameter to @escaping where i will pass the TodoItem model asyncronously

'TodoFeedViewModel' changes:
    - Introduced an extension for the Array to map an array of 'TodoItem' into TodoItemViewModel'
    - defined loadFeed closure passing it via the init where adapter.load() metod is triggered, viewmodel calls load in the adapter via the closure and initalizer with the assigned function adapter.load from the composition
    - Introduced a mechanism for storage and fast lookup for original item such [TodoItem] with the ThreadSafe dictionary where I store UUID of the task adn a value as its index in the array of [TodoItem]

Fenominall committed:
Added explonation comment to the 'ThreadSafeDictionary' why I do not need weak self in the isolated withing the queue.async block
            // ThreadSafeArray owns the queue, and the queue does not outlive ThreadSafeArray
            // Weak self not needed

Fenominall committed:
feat: Implement thread-safe array with concurrent access handling

Fenominall committed:
Fixed typo, renamed the ThreadSafeArray

Fenominall committed:
Updated implemtnation fo the addOrUpdateTasks func:
    - using update method for tasks update
    - extracted getting the count of the originalItems in aseparate func without neededing to use -1 for getting the correct index

Fenominall committed:
Introduced selectItem method in the TodoFeedViewModel for selection a item by id from the TaskListView

Fenominall committed:
Intoduced TodoFeedStore with the two NSmanageObjects
ManagedTodoItem that will store all the properties and ManagedCache that will store an array in NSOrderedSet to guarantee taht we have ordered collection

Fenominall committed:
Introduced helper variabled for converting an list of ManagedTodoitem into LocalTodoItem

Fenominall committed:
Fixy type for  var lcoal in ManagedTodoItem

Fenominall committed:
Reffactored perforam async func to make it compitable with the async await

Provided implemantation for the retrieve method in the CoreDataFeedStore for providing stored data

Fenominall committed:
Updated TodoItem and its DTOs to ahve non optional URL as it should containt a url toan image
- Introduced CoreDataFeedStoreError with teh possible error for CRUD operations in CoreData

Fenominall committed:
Did make imageURL non optional for the ManagedTodoItem and LocalTodoItem

Fenominall committed:
Implement CoreData caching for TodoItems with performance optimizations

- Added ManagedTodoItem entity and extended its functionality with methods for data retrieval, batch insertion, and cache updating.
- Introduced `userInfo` caching in the NSManagedObjectContext for efficient image data handling.
- Optimized cache management by using `userInfo` to store and retrieve image data, reducing unnecessary fetches.
- Implemented `clearUserInfo(in:)` for clearing cached data after insertions or deletions to improve memory management.
- Added functionality to update, insert, and fetch `ManagedCache` efficiently with new tasks.
- Updated `CoreDataFeedStore` to handle asynchronous retrieval, insertion, and updating of TodoItems.

Fenominall committed:
Provided implementation for the TodoFeedImageStore in CoreDataFeedStore to store and retrieve binary data from teh CoreData

vadim-chistiakov commented:
Посмотри что такое .gitignore и как с ним работать, будет полезно

vadim-chistiakov commented on ProtonTodoDevTest/ProtonTodoDevTest/Shared API/URLSessionHTTPClient.swift:
```diff
@@ -0,0 +1,28 @@
+        self.session = session
+    }
+    
+    private struct UnexpectedValueRepresentaion: Error {}
```
Как правило делают enum c разными ошибками, так будет удобнее

vadim-chistiakov commented on ProtonTodoDevTest/ProtonTodoDevTestiOS/SharedHelpers/ConvertingExtensions.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+func iso8601FormattedString(from date: Date) -> String {
```
такие функции как правило в экстеншинах у String or Date прописываются.

vadim-chistiakov commented on ProtonTodoDevTest/ProtonTodoDevTestiOS/Todo Details UI/Views/TaskDetailView.swift:
```diff
@@ -0,0 +1,44 @@
+
+import SwiftUI
+
+public struct TaskDetailView: View {
```
тебе не обязательно делать все public пока у тебя один модуль в проекте

vadim-chistiakov commented on ProtonTodoDevTest/ProtonTodoDevTestiOS/Todo Feed UI/Presentation/ViewModels/TodoItemViewModel.swift:
```diff
@@ -0,0 +1,56 @@
+import Foundation
+import ProtonTodoDevTest
+
+public struct TodoItemViewModel: Identifiable, Hashable {
```
Если это просто модель с данными, она не должна называться ViewModel

vadim-chistiakov commented on ProtonTodoDevTest/ProtonTodoDevTestiOS/Todo Feed UI/Views/TaskListViews/TaskRowViewComponents/TaskRowSeparator.swift:
```diff
@@ -0,0 +1,20 @@
+import SwiftUI
+
+// TaskSeparator - Reusable separator line under the task row
+struct TaskRowSeparator: View {
```
Усложнил тут. Чтобы добавить сепаратор, достаточно стиль у таблицы выбрать с сепаратором

vadim-chistiakov commented on ProtonTodoDevTest/ProtonTodoDevTestiOS/Todo Feed UI/Views/TaskListViews/TaskRowViewComponents/TaskRowView.swift:
```diff
@@ -0,0 +1,46 @@
+    }
+}
+
+//TaskRowInfoVIew(
```
закоментированный код на ревью не принято отправлять. Используй Preview для отладки

vadim-chistiakov commented:
Поменьше чата жпт. Старайся сам придумывать решения проблем. К чату обращайся когда застрял

Fenominall committed:
Added some description for clarification

Fenominall committed:
Added extension on Date for converting Date into String with iso8601 Format

Fenominall committed:
Provided implementation for the HTTP response errors in the URLSessionHTTPClient
Covering cases 200..<300, 400..<500, 500..<600

Fenominall committed:
Did rename 'TodoItemViewModel' DTO to the 'TodoItemPresentationModel' to emphasze that the DTo is destinated for the UILayer
A simple model should not have appendix ViewModel

Fenominall committed:
Did rename the converting method from TodoItem to TodoItemPresenationModel to the name mapToViewRepresentationModel to clarify the intent
Changed the privacy level to internal cause there is no need to expose it to another module

Fenominall committed:
Introduced onCompletionStatusChange closures in the views in order to provide todo status change from the 'TaskRowCompletionToggleView' > 'TaskRowView' > 'TaskListView' > 'TodoFeedViewModel' to update the todo in the locale storage

- Defined todoStatusChaged closure that receives a model on the status change completion toggle, providing the toggled TodoItem via the init
- Defined updateTodoStatus method that received a status change as Bool and a todo id on which the toggle happend, updating the model and passing it via todoStatusChaged closure

Refactored TaskListViewComposer:
    - Added TodoItemSaveable type to the composedViewWith method and extension for the method with name cachingItem to be able to save updated   single item to the local storage

-Fixed the indentetion for the helpers methods in the entry point of the app in ProtonTodoDevTestAppApp file
- Passing localTodoFeedManager as a parameter for the todoItemSaveable type in TaskListViewComposer for the composedViewWith method

Fenominall committed:
Added helper comment for the Combine helpers

Fenominall committed:
Removed unused files

Fenominall committed:
Fixed conditional logic to display TaskDetailImageView only when DownloadImageButton is tapped and tiggered isImageDownloaded property in the viewModel

if isImageDownloaded is true providing an image data to the TaskDetailImageView, checking that image data is not empty and uwrapping the data
else  showing Progress view

Fenominall committed:
Did renamed TaskRowCompletionToggle to TaskRowCompletionToggleView to clarify the intent for the view

Fenominall committed:
Refactored 'ThreadSafeArra' and 'ThreadSafeDictionary' to support async/await instead of GCD using actor propogates thread safe isolated environment.
Refactored key methods in the 'TodoFeedViewModel' such addOrUpdateTasks, selectItem, selectItem to support async async await for use with  'ThreadSafeArra' and 'ThreadSafeDictionary'.
Wrapping async blocks in the Task in the viewModel, when Task added inside an ObservableObject, it doesn’t strongly retain self.
Refactored completion closures in the view to support async await manner and for onCompletionStatusChange closure to provide the value to the viewmodel

Fenominall committed:
Removed unnecessart code when adding new todos with data, cause if the data is mpty initalty we do not insert it

Fenominall committed:
Refactored receive dispatch for the combine abstraction helper, introduced MockHTTPClient with mock data for data loading simulation

Fenominall committed:
Fixed label name typo in the ContenvtView for Upcoming tasks view tab

Fenominall committed:
Added imageURL property to the TodoItemPresntationModel DTO for providing imageURl for image downlaoding

Fenominall committed:
FIxed indentation

Fenominall committed:
Refactored:
    - Did add a closure in the TaskListView that gets a TodoItemPresentationModel and returns TaskRowView, this was done fr abstaction cause the TaskRowView will communicate with introduced TaskRoewViewModel taht will be downloading the image for each TaskRow and updating the todo status change
- Remove the updatng status logic from the TodoFeedViewModel
- Introduced TaskRowViewModel - it receives TodoItemPresentationModel, when the TaskRowView is loaded the image is downloanded and provided to the view for data redndering view task published property
- Changed the view models properties type in the TaskListView and TaskRowView to not recreate a view model each time its reloaded
 - Populating recived model data for the views in the TaskRowView
- Introduced ImageDataLoadingPresentationAdapter as a LoadResourcePresentationAdapter that loads data in the 'TaskListViewComposer'
- Created a factory helper method in the 'TaskListViewComposer' that created TaskRowView for each received model and downloads an image for it

Fenominall committed:
Removed unused imageData propert from the 'TaskRowViewModel'

Fenominall committed:
Introduced 'AsyncimageView' that received data and asyncronously converts it into the UIImage then assiging to the Image view otherwise processing, shwoing ProgressView until finished
Using the introduced 'AsyncImageView' in the 'TaskRoewImageView' for image rendering with the received data

Fenominall committed:
Fixed indentation for the preview

Fenominall committed:
Changed the privacy level to public for createdAt and dueDate properties in the TodoItemPresentationModel to be able to convert it into the totoItem

Fenominall committed:
Handeling TodoItem todo status update by saving the currect Bool statuse in the local storage when changed
- Introduced a new property receivedTask for updating the status in the local storage of TodoItem  and reanmed the @Published task to publishedTask for updating the UI

Fenominall committed:
Handling tasks filtering for the two tabs with All Tasks and Upcoming tasks reusing TaskListView
Introduced a closure that provides a predicate in the composition level and that being applied inside the 'TodoViewModel' for tasks delivering based on the predicate assgning to the @Piblishes tasks property

Fenominall committed:
Introduced AlertViewModifier with the alertView func defined in the View extension in order to present an alert when an eror occurs
handeling the error case to show an error in the TaskListView
Defined isShowingError and Error Bindable properties in the view to handle the error case  for the alertview modifier

Fenominall committed:
Removed the some view modifiers from the AsyncImageView to be able to set them on the view instantiantion

Fenominall committed:
Fixed preview error in the TaskDetailView for the given imageURL

Fenominall committed:
Using AsyncImageView in the TaskDetailImageView

Fenominall committed:
Refactored TaskRowImageView for the AsyncImageView by adding necessary view modifiers

Fenominall committed:
Defined TodoDetailViewComposer for composing TodoItemkDetailView

- Implemented the image loading logic inside the TodoItemDetailViewModel handeling the loading and image data providing triggering the action from the
- Did rename DownloadImageButtonView to TodoDetailImageView

Fenominall committed:
Fixed typo

Fenominall committed:
Reanmed view

Fenominall committed:
changed the order of the loading state to finish loading state before just before the data is provided

Fenominall committed:
Did rename the object to Todo prefix to clarify the intent of the app core model as TodoItem

Fenominall committed:
Did renamed initalizer properties of the TodoListView for the todoRowView

Fenominall committed:
feature: Encrypt Core Data storage using CryptoKit

- Integrated CryptoKit to encrypt and decrypt todo data before storing in Core Data.
- Introduced CryptoManager for storing and retrieving a secure SymmetricKey for CryptoKit
- Added helper methods for secure encryption and decryption using CryptoKit
